src/main/java/org/prebid/server/bidder/thetradedesk/TheTradeDeskBidder.java

```diff
@@ -0,0 +1,206 @@
+                        ? StringUtils.isNotBlank(extImp.getPublisherId())
+                        ? extImp.getPublisherId()
+                        : publisherId
+                        : publisherId;
+
+                modifiedImps.add(modifyImp(imp));
+            } catch (PreBidException e) {
+                return Result.withError(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        final BidRequest outgoingRequest = modifyRequest(request, modifiedImps, publisherId);
+        final HttpRequest<BidRequest> httpRequest = BidderUtil.defaultRequest(
+                outgoingRequest,
+                makeHeaders(),
+                resolveEndpoint(),
+                mapper);
+
+        return Result.withValue(httpRequest);
+    }
+
+    private ExtImpTheTradeDesk parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+    }
+
+    private Imp modifyImp(Imp imp) {
+        final Banner banner = imp.getBanner();
+
+        if (banner != null && CollectionUtils.isNotEmpty(banner.getFormat())) {
+            final Format format = banner.getFormat().getFirst();
+            return imp.toBuilder()
+                    .banner(banner.toBuilder().w(format.getW()).h(format.getH()).build())
+                    .build();
+        }
+
+        return imp;
+    }
+
+    private BidRequest modifyRequest(BidRequest request, List<Imp> modifiedImps, String publisherId) {
+        return request.toBuilder()
+                .imp(modifiedImps)
+                .site(modifySite(request, publisherId))
+                .app(modifyApp(request, publisherId))
+                .build();
+    }
+
+    private Site modifySite(BidRequest request, String publisherId) {
+        final Site site = request.getSite();
+        if (site == null) {
+            return null;
+        }
+
+        return site.toBuilder()
+                .publisher(modifyPublisher(site.getPublisher(), publisherId))
+                .build();
+    }
+
+    private static Publisher modifyPublisher(Publisher publisher, String publisherId) {
+        if (publisher == null) {
+            return Publisher.builder().id(publisherId).build();
+        }
+
+        return publisher.toBuilder()
+                .id(StringUtils.isNotBlank(publisherId) ? publisherId : publisher.getId())
+                .build();
+    }
+
+    private App modifyApp(BidRequest request, String publisherId) {
+        final Site site = request.getSite();
+        final App app = request.getApp();
+
+        if (site != null) {
+            return app;
+        }
+
+        if (app == null) {
+            return null;
+        }
+
+        return app.toBuilder()
+                .publisher(modifyPublisher(app.getPublisher(), publisherId))
+                .build();
+    }
+
+    private MultiMap makeHeaders() {
```

Comment:
In this case headers are always same, so we can have them on class level, to not create each time